README.md
# blog_ippon_sonarqube_jenkinsfile
Code source du tuto Jenkinsfile with Sonarqube !


CodeNarc-0.25.2_blog/CHANGELOG.txt

#TODO: Sort the following line into the file
- CheckUserInputInNode rule (jenkinsfile) - Check that there is no 'input' statement in a 'node' block.

CodeNarc Change Log
-------------------------------------------------------------------------------
http://www.codenarc.org

Changes in version 0.25.2 (Mar 2016)
--------------------------------------
- #150: BracesForIfElseRule can't process enum files.
- #152: VariableNameRule: Variables with annotations can cause ArrayIndexOutOfBoundsException.
- #149: BracesForForLoop, BracesForIfElseRule: Improve validation if there are GString curly braces as part of expression.
        BracesForTryCatchFinally: Improve validation if there are curly braces in comments.


Changes in version 0.25.1 (Feb 2016)
--------------------------------------
- #147: UnusedImport: misses import reference within GString.
- Change NoTabCharacter rule priority to 3.


Changes in version 0.25.0 (Feb 2016)
--------------------------------------
NEW RULES
- #135: NoTabCharacter rule (convention) - Check that all source files do not contain the tab character (Yuriy Chulovskyy)
- #136: TrailingComma rule (convention). (Yuriy Chulovskyy)

UPDATED/ENHANCED RULES AND BUG FIXES
- #146: Validate regular expressions when setting rule properties applyToFilesMatching and doNotApplyToFilesMatching.
- #145: Reduce number of concurrent threads to number of available processors.
- #144: LineLength rule: Add ignoreLineRegex.
- #143: EmptyClass rule: also ignore abstract classes.
- #124: UnusedVariable rule: Ignore variables annotated with @BaseScript.
- #141: UnnecessaryDefInFieldDeclaration rule: false violation if comment includes the word "def".
- #139: IdeTextReportWriter: Default line number to 0 so that IDE links are valid.
- #131: UnusedImport rule does not handle Scala. Fix for #115. (Yuriy Chulovskyy)
- #133: Updated documentation for BracesForIfElse rule. (Yuriy Chulovskyy)
- #130: Fixed issue #119 BracesForIfElse with curly braces inside GString. (Yuriy Chulovskyy)
- #129: Fixed issue #121 False positive for UnusedPrivateField. (Yuriy Chulovskyy)
- #127: Fixed issue #126 False positive UnnecessaryPublicModifier for groovy (Yuriy Chulovskyy)
- #122: Fix typo: contant -> constant (Dominik Przybysz)


Changes in version 0.24.1 (Aug 2015)
--------------------------------------
- #114: Excluding baseline violations (excludeBaseline) does not match/ignore all baseline violations.
    * Fixed BaselineResultsProcessor to properly handle encoded messages.
    * Fixed BaselineResultsProcessor to properly filter out multiple violations of the same rule per file.
    * Fixed BaselineResultsProcessor to properly handle violations with null/empty messages.
    * Added logging when loading baseline violations file and count of ignored violations.
- #117: Enums should not trigger check for SerialVersionUID.


Changes in version 0.24 (July 2015)
--------------------------------------
NEW RULES
- #110: AssignmentToStaticFieldFromInstanceMethod rule (design) - Checks for assignment to a static field from an instance method.
- #111: ClassNameSameAsSuperclass rule (naming) - Checks for any class that has an identical name to its superclass, other than the package. This can be very confusing.
- #111: InterfaceNameSameAsSuperInterface rule (naming) - Checks for any interface that has an identical name to its super-interface, other than the package. This can be very confusing.

UPDATED/ENHANCED RULES AND BUG FIXES
- #91: ClassJavadoc rule: Allow regular comments or annotations before Javadoc section. (Thanks to Sébastien Launay)
- #90: UnnecessarySafeNavigationOperator rule: Fix false positive on spread operator.
- #95: Some tests incompatible with Java 8. (Thanks to Artur Gajowy)
- #107: FileEndsWithoutNewlineRule to handle empty source. (Thanks to René Scheibe)
- #108: Fix NPE when analyzing file with trait; fix #93. (Thanks to Dominik Przybysz)

FRAMEWORK AND INFRASTRUCTURE
- #92: New SortableHtmlReportWriter, report type "sortable". Sort by rules (with the most violations) and files (with the most violations), as well as rule name and rule priority.
- #89: Provide mechanism for excluding a baseline violations file. Add "baseline" report type and associated "excludeBaseline" property to Ant Task.
- #39: Convert project to Gradle. (Thanks very much to Tomasz Przybysz, Marcin Erdmann and Kyle Boon)
- #100: Remove IDE project files from source control. Use Gradle IDEA Plugin.
- #106: HtmlReportWriter add includeRuleDescriptions property.
- #109: Enable @SuppressWarnings to work on constructors.
- #112: Fix file encoding for CHANGELOG.txt in codenarc.groovy command-line script


Changes in version 0.23 (Feb 2015)
--------------------------------------
NEW RULES
- #73: NestedForLoop rule (design) - Checks for nested for loops. (Thanks to Maciej Ziarko)
- #80: ParameterCount rule (size) - Checks if the number of parameters in method/constructor exceeds the number of parameters specified by the maxParameters property. (Thanks to Maciej Ziarko)

UPDATED/ENHANCED RULES AND BUG FIXES
- #69: PrivateFieldCouldBeFinal rule support for JPA entities. (Thanks to Maciej Ziarko)
- #70: GStringExpressionWithinString rule should ignore quasi GString expressions inside annotations. (Thanks to Maciej Ziarko)
- #76: SpaceAfterClosingBrace and SpaceBeforeOpeningBrace rules: GStrings with embedded closures incorrectly cause violations. (Thanks to Maciej Ziarko)
- #59: SpaceAfterClosingBrace false violation if string contains closing brace. (fixed by #76)
- #79: JUnitTestMethodWithoutAssert support for JUnit's ExpectedException way to verify exceptions. (Thanks to Maciej Ziarko)
- #66: SpaceAfterComma rule doesn't check constructor calls or constructor declarations.
- #82: IdeTextReportWriter violation link (loc=..) does not work in Eclipse.
- #81: XML Report: Escape illegal XML characters within the source line or message.

sonar-groovy-master_blog/sonar-groovy-plugin/src/test/java/org/sonar/plugins/groovy/GroovySensorTest.java
/*
 * Sonar Groovy Plugin
 * Copyright (C) 2010-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plugins.groovy;

import org.junit.Test;
import org.mockito.Matchers;
import org.mockito.Mockito;
import org.sonar.api.batch.fs.internal.DefaultFileSystem;
import org.sonar.api.batch.fs.internal.DefaultInputDir;
import org.sonar.api.batch.fs.internal.DefaultInputFile;
import org.sonar.api.batch.sensor.internal.DefaultSensorDescriptor;
import org.sonar.api.batch.sensor.internal.SensorContextTester;
import org.sonar.api.config.Settings;
import org.sonar.api.measures.CoreMetrics;
import org.sonar.api.measures.FileLinesContext;
import org.sonar.api.measures.FileLinesContextFactory;
import org.sonar.plugins.groovy.foundation.Groovy;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class GroovySensorTest {

  private Settings settings = new Settings();
  private FileLinesContextFactory fileLinesContextFactory = mock(FileLinesContextFactory.class);
  private DefaultFileSystem fileSystem = new DefaultFileSystem(new File("."));
  private GroovySensor sensor = new GroovySensor(settings, fileLinesContextFactory, fileSystem);

  @Test
  public void do_nothing_when_no_groovy_file() throws IOException {
    SensorContextTester context = SensorContextTester.create(new File(""));
    context = Mockito.spy(context);
    sensor = new GroovySensor(settings, fileLinesContextFactory, context.fileSystem());
    sensor.execute(context);

    Mockito.verify(context, Mockito.never()).newHighlighting();
  }

  @Test
  public void compute_metrics() throws IOException {
    testMetrics(false, 5);
  }

  @Test
  public void compute_metrics_ignoring_header_comment() throws IOException {
    testMetrics(true, 3);
  }

  private void testMetrics(boolean headerComment, int expectedCommentMetric) throws IOException {
    settings.appendProperty(GroovyPlugin.IGNORE_HEADER_COMMENTS, "" + headerComment);
    File sourceDir = new File("src/test/resources/org/sonar/plugins/groovy/gmetrics");
    SensorContextTester context = SensorContextTester.create(new File(""));

    File sourceFile = new File(sourceDir, "Greeting.groovy");
    fileSystem = context.fileSystem();
    fileSystem.add(new DefaultInputDir("", sourceDir.getPath()));
    DefaultInputFile groovyFile = new DefaultInputFile("", sourceFile.getPath())
      .setLanguage(Groovy.KEY)
      .initMetadata(new String(Files.readAllBytes(sourceFile.toPath()), "UTF-8"));
    fileSystem.add(groovyFile);
    fileSystem.add(new DefaultInputFile("", "unknownFile.groovy").setLanguage(Groovy.KEY));

    FileLinesContext fileLinesContext = mock(FileLinesContext.class);
    when(fileLinesContextFactory.createFor(any(DefaultInputFile.class))).thenReturn(fileLinesContext);

    sensor = new GroovySensor(settings, fileLinesContextFactory, fileSystem);
    sensor.execute(context);

    String key = groovyFile.key();
    assertThat(context.measure(key, CoreMetrics.FILES).value()).isEqualTo(1);
    assertThat(context.measure(key, CoreMetrics.CLASSES).value()).isEqualTo(2);
    assertThat(context.measure(key, CoreMetrics.FUNCTIONS).value()).isEqualTo(2);

    assertThat(context.measure(key, CoreMetrics.LINES).value()).isEqualTo(33);
    assertThat(context.measure(key, CoreMetrics.NCLOC).value()).isEqualTo(17);
    assertThat(context.measure(key, CoreMetrics.COMMENT_LINES).value()).isEqualTo(expectedCommentMetric);

    assertThat(context.measure(key, CoreMetrics.COMPLEXITY).value()).isEqualTo(4);
    assertThat(context.measure(key, CoreMetrics.COMPLEXITY_IN_CLASSES).value()).isEqualTo(4);
    assertThat(context.measure(key, CoreMetrics.COMPLEXITY_IN_FUNCTIONS).value()).isEqualTo(4);

    assertThat(context.measure(key, CoreMetrics.FUNCTION_COMPLEXITY_DISTRIBUTION).value()).isEqualTo("1=0;2=2;4=0;6=0;8=0;10=0;12=0");
    assertThat(context.measure(key, CoreMetrics.FILE_COMPLEXITY_DISTRIBUTION).value()).isEqualTo("0=1;5=0;10=0;20=0;30=0;60=0;90=0");

    // 11 times for comment because we register comment even when ignoring header comment
    Mockito.verify(fileLinesContext, Mockito.times(11)).setIntValue(Mockito.eq(CoreMetrics.COMMENT_LINES_DATA_KEY), Matchers.anyInt(), Mockito.eq(1));
    Mockito.verify(fileLinesContext, Mockito.times(17)).setIntValue(Mockito.eq(CoreMetrics.NCLOC_DATA_KEY), Matchers.anyInt(), Mockito.eq(1));
    Mockito.verify(fileLinesContext).setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, 18, 1);
    Mockito.verify(fileLinesContext).setIntValue(CoreMetrics.NCLOC_DATA_KEY, 18, 1);
    // Only "Greeting.groovy" is part of the file system.
    Mockito.verify(fileLinesContext, Mockito.times(1)).save();
  }

  @Test
  public void compute_coupling_metrics() throws IOException {
    SensorContextTester context = SensorContextTester.create(new File(""));

    fileSystem = context.fileSystem();
    // package 'org' contains class 'Greeting', used by no other class, but using classes 'Bar' and 'Foo'
    DefaultInputDir org = addFileWithParentFolder("src/test/resources/org/sonar/plugins/groovy/gmetricswithcoupling/org", "Greeting.groovy");
    // package 'org.foo' contains class 'Foo', used by class 'Greeting', and using class 'Bar'
    DefaultInputDir org_foo = addFileWithParentFolder("src/test/resources/org/sonar/plugins/groovy/gmetricswithcoupling/org/foo", "Foo.groovy");
    // package 'org.bar' contains class 'Bar', used by classes 'Greeting' and 'Foo', but using no other class
    DefaultInputDir org_bar = addFileWithParentFolder("src/test/resources/org/sonar/plugins/groovy/gmetricswithcoupling/org/bar", "Bar.groovy");

    FileLinesContext fileLinesContext = mock(FileLinesContext.class);
    when(fileLinesContextFactory.createFor(any(DefaultInputFile.class))).thenReturn(fileLinesContext);

    sensor = new GroovySensor(settings, fileLinesContextFactory, fileSystem);
    sensor.execute(context);

    assertCouplingMeasureAre(context, org.key(), 3, 1.0, 3, 1.0);
    assertCouplingMeasureAre(context, org_foo.key(), 1, 1.0, 1, 1.0);
    assertCouplingMeasureAre(context, org_bar.key(), 2, 2.0, 0, 0.0);
  }

  private static void assertCouplingMeasureAre(SensorContextTester context, String key, Object afferentTot, Object afferentAvg, Object efferentTot, Object efferentAvg) {
    assertThat(context.measure(key, GroovyMetrics.AFFERENT_COUPLING_TOTAL.key()).value()).isEqualTo(afferentTot);
    assertThat(context.measure(key, GroovyMetrics.AFFERENT_COUPLING_AVERAGE.key()).value()).isEqualTo(afferentAvg);

    assertThat(context.measure(key, GroovyMetrics.EFFERENT_COUPLING_TOTAL.key()).value()).isEqualTo(efferentTot);
    assertThat(context.measure(key, GroovyMetrics.EFFERENT_COUPLING_AVERAGE.key()).value()).isEqualTo(efferentAvg);
  }

  private DefaultInputDir addFileWithParentFolder(String dirPath, String fileName) throws IOException {
    File dir = new File(dirPath);
    File file = new File(dir, fileName);
    DefaultInputDir inputDir = new DefaultInputDir("", dir.getPath());
    fileSystem.add(inputDir);
    fileSystem.add(new DefaultInputFile("", file.getPath())
      .setLanguage(Groovy.KEY)
      .initMetadata(new String(Files.readAllBytes(file.toPath()), "UTF-8")));
    return inputDir;
  }

  @Test
  public void test_toString() {
    assertThat(sensor.toString()).isEqualTo("GroovySensor");
  }

  @Test
  public void test_description() {
    DefaultSensorDescriptor defaultSensorDescriptor = new DefaultSensorDescriptor();
    sensor.describe(defaultSensorDescriptor);
    assertThat(defaultSensorDescriptor.languages()).contains();
  }

}


sonar-groovy-master_blog/sonar-groovy-plugin/src/test/java/org/sonar/plugins/groovy/codenarc/CodeNarcRulesDefinitionTest.java
/*
 * Sonar Groovy Plugin
 * Copyright (C) 2010-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plugins.groovy.codenarc;

import java.util.LinkedList;
import java.util.List;
import org.junit.Test;
import org.sonar.api.server.rule.RulesDefinition;
import org.sonar.api.server.rule.RulesDefinition.Rule;
import org.sonar.plugins.groovy.foundation.Groovy;

import static org.assertj.core.api.Assertions.assertThat;

public class CodeNarcRulesDefinitionTest {

  @Test
  public void test() {
    CodeNarcRulesDefinition definition = new CodeNarcRulesDefinition();
    RulesDefinition.Context context = new RulesDefinition.Context();
    definition.define(context);
    RulesDefinition.Repository repository = context.repository(CodeNarcRulesDefinition.REPOSITORY_KEY);

    assertThat(repository.name()).isEqualTo(CodeNarcRulesDefinition.REPOSITORY_NAME);
    assertThat(repository.language()).isEqualTo(Groovy.KEY);

    List<Rule> rules = repository.rules();
    assertThat(rules).hasSize(348);

    List<String> missingDebt = new LinkedList<>();
    for (Rule rule : rules) {
      assertThat(rule.key()).isNotNull();
      assertThat(rule.internalKey()).isNotNull();
      assertThat(rule.name()).isNotNull();
      assertThat(rule.htmlDescription()).isNotNull();
      if (rule.debtRemediationFunction() == null) {
        missingDebt.add(rule.key());
      }
    }
    // From SONARGROOV-36, 'org.codenarc.rule.generic.IllegalSubclassRule' does not have debt by purpose
    assertThat(missingDebt).containsOnly("org.codenarc.rule.generic.IllegalSubclassRule.fixed");

    Rule rule = repository.rule("org.codenarc.rule.braces.ElseBlockBracesRule");
    assertThat(rule.params()).hasSize(1);
    assertThat(rule.params().get(0).defaultValue()).isEqualToIgnoringCase("false");
  }
}


sonar-groovy-master_blog/sonar-groovy-plugin/src/main/java/org/sonar/plugins/groovy/GroovySensor.java
/*
 * Sonar Groovy Plugin
 * Copyright (C) 2010-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.plugins.groovy;

import groovyjarjarantlr.Token;
import groovyjarjarantlr.TokenStream;
import groovyjarjarantlr.TokenStreamException;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Serializable;
import java.math.BigDecimal;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import org.apache.commons.io.FileUtils;
import org.codehaus.groovy.antlr.GroovySourceToken;
import org.codehaus.groovy.antlr.parser.GroovyLexer;
import org.codehaus.groovy.antlr.parser.GroovyTokenTypes;
import org.gmetrics.result.MetricResult;
import org.gmetrics.result.MutableMapMetricResult;
import org.gmetrics.result.NumberMetricResult;
import org.gmetrics.result.SingleNumberMetricResult;
import org.gmetrics.resultsnode.ClassResultsNode;
import org.gmetrics.resultsnode.PackageResultsNode;
import org.gmetrics.resultsnode.ResultsNode;
import org.sonar.api.batch.fs.FileSystem;
import org.sonar.api.batch.fs.InputComponent;
import org.sonar.api.batch.fs.InputDir;
import org.sonar.api.batch.fs.InputFile;
import org.sonar.api.batch.measure.Metric;
import org.sonar.api.batch.sensor.Sensor;
import org.sonar.api.batch.sensor.SensorContext;
import org.sonar.api.batch.sensor.SensorDescriptor;
import org.sonar.api.ce.measure.RangeDistributionBuilder;
import org.sonar.api.config.Settings;
import org.sonar.api.measures.CoreMetrics;
import org.sonar.api.measures.FileLinesContext;
import org.sonar.api.measures.FileLinesContextFactory;
import org.sonar.api.utils.log.Logger;
import org.sonar.api.utils.log.Loggers;
import org.sonar.plugins.groovy.foundation.Groovy;
import org.sonar.plugins.groovy.foundation.GroovyFileSystem;
import org.sonar.plugins.groovy.foundation.GroovyHighlighterAndTokenizer;
import org.sonar.plugins.groovy.gmetrics.GMetricsSourceAnalyzer;

public class GroovySensor implements Sensor {
  private static final Logger LOG = Loggers.get(GroovySensor.class);

  private static final String CYCLOMATIC_COMPLEXITY_METRIC_NAME = "CyclomaticComplexity";
  private static final String EFFERENT_COUPLING_METRIC_NAME = "EfferentCoupling";
  private static final String AFFERENT_COUPLING_METRIC_NAME = "AfferentCoupling";

  private static final Number[] FUNCTIONS_DISTRIB_BOTTOM_LIMITS = {1, 2, 4, 6, 8, 10, 12};
  private static final Number[] FILES_DISTRIB_BOTTOM_LIMITS = {0, 5, 10, 20, 30, 60, 90};

  private static final Set<String> EMPTY_COMMENT_LINES = Arrays.stream(new String[] {"/**", "/*", "*", "*/", "//"}).collect(Collectors.toSet());

  private final Settings settings;
  private final FileLinesContextFactory fileLinesContextFactory;
  private final GroovyFileSystem groovyFileSystem;

  private int loc = 0;
  private int comments = 0;
  private int currentLine = 0;
  private FileLinesContext fileLinesContext;

  public GroovySensor(Settings settings, FileLinesContextFactory fileLinesContextFactory, FileSystem fileSystem) {
    this.settings = settings;
    this.fileLinesContextFactory = fileLinesContextFactory;
    this.groovyFileSystem = new GroovyFileSystem(fileSystem);
  }

  @Override
  public void describe(SensorDescriptor descriptor) {
    descriptor./*onlyOnLanguage(Groovy.KEY).*/name(this.toString());
  }

  @Override
  public void execute(SensorContext context) {
    if (groovyFileSystem.hasGroovyFiles()) {
      List<InputFile> inputFiles = groovyFileSystem.sourceInputFiles();
      computeBaseMetrics(context, inputFiles);
      computeGroovyMetrics(context, inputFiles);
      highlightFiles(context, groovyFileSystem.groovyInputFiles());
    }
  }

  private static void computeGroovyMetrics(SensorContext context, List<InputFile> inputFiles) {
    GMetricsSourceAnalyzer metricsAnalyzer = new GMetricsSourceAnalyzer(context.fileSystem(), inputFiles);

    metricsAnalyzer.analyze();

    for (Entry<InputFile, List<ClassResultsNode>> entry : metricsAnalyzer.resultsByFile().entrySet()) {
      processFile(context, entry.getKey(), entry.getValue());
    }

    for (Entry<InputDir, PackageResultsNode> entry : metricsAnalyzer.resultsByPackage().entrySet()) {
      processPackage(context, entry.getKey(), entry.getValue().getMetricResults());
    }
  }

  private static void processFile(SensorContext context, InputFile sonarFile, Collection<ClassResultsNode> results) {
    int classes = 0;
    int methods = 0;
    int complexity = 0;
    int complexityInFunctions = 0;

    RangeDistributionBuilder functionsComplexityDistribution = new RangeDistributionBuilder(FUNCTIONS_DISTRIB_BOTTOM_LIMITS);

    for (ClassResultsNode result : results) {
      classes += 1;

      for (ResultsNode resultsNode : result.getChildren().values()) {
        methods += 1;
        Optional<MetricResult> cyclomaticComplexity = getCyclomaticComplexity(resultsNode.getMetricResults());
        if (cyclomaticComplexity.isPresent()) {
          int value = (Integer) ((SingleNumberMetricResult) cyclomaticComplexity.get()).getNumber();
          functionsComplexityDistribution.add(value);
          complexityInFunctions += value;
        }
      }

      Optional<MetricResult> cyclomaticComplexity = getCyclomaticComplexity(result.getMetricResults());
      if (cyclomaticComplexity.isPresent()) {
        int value = (Integer) ((NumberMetricResult) cyclomaticComplexity.get()).getValues().get("total");
        complexity += value;
      }
    }

    saveMetric(context, sonarFile, CoreMetrics.FILES, 1);
    saveMetric(context, sonarFile, CoreMetrics.CLASSES, classes);
    saveMetric(context, sonarFile, CoreMetrics.FUNCTIONS, methods);
    saveMetric(context, sonarFile, CoreMetrics.COMPLEXITY, complexity);
    saveMetric(context, sonarFile, CoreMetrics.COMPLEXITY_IN_CLASSES, complexity);
    saveMetric(context, sonarFile, CoreMetrics.COMPLEXITY_IN_FUNCTIONS, complexityInFunctions);
    saveMetric(context, sonarFile, CoreMetrics.FUNCTION_COMPLEXITY_DISTRIBUTION, functionsComplexityDistribution.build());

    RangeDistributionBuilder fileComplexityDistribution = new RangeDistributionBuilder(FILES_DISTRIB_BOTTOM_LIMITS);
    fileComplexityDistribution.add(complexity);
    saveMetric(context, sonarFile, CoreMetrics.FILE_COMPLEXITY_DISTRIBUTION, fileComplexityDistribution.build());
  }

  private static Optional<MetricResult> getCyclomaticComplexity(List<MetricResult> metricResults) {
    return metricResults
      .stream()
      .filter(metricResult -> CYCLOMATIC_COMPLEXITY_METRIC_NAME.equals(metricResult.getMetric().getName()))
      .findAny();
  }

  private static void processPackage(SensorContext context, InputDir inputDir, List<MetricResult> metricResults) {
    for (MetricResult metricResult : metricResults) {
      org.gmetrics.metric.Metric metric = metricResult.getMetric();
      String metricName = metric.getName();
      if (EFFERENT_COUPLING_METRIC_NAME.equals(metricName)) {
        MutableMapMetricResult result = (MutableMapMetricResult) metricResult;
        saveMetric(context, inputDir, GroovyMetrics.EFFERENT_COUPLING_TOTAL, getTotalValue(result));
        saveMetric(context, inputDir, GroovyMetrics.EFFERENT_COUPLING_AVERAGE, getAverageValue(result));
      } else if (AFFERENT_COUPLING_METRIC_NAME.equals(metricName)) {
        MutableMapMetricResult result = (MutableMapMetricResult) metricResult;
        saveMetric(context, inputDir, GroovyMetrics.AFFERENT_COUPLING_TOTAL, getTotalValue(result));
        saveMetric(context, inputDir, GroovyMetrics.AFFERENT_COUPLING_AVERAGE, getAverageValue(result));
      }
    }
  }

  private static Integer getTotalValue(MutableMapMetricResult result) {
    return (Integer) result.getAt("total");
  }

  private static double getAverageValue(MutableMapMetricResult result) {
    Object avg = result.getAt("average");
    BigDecimal avgValue = (avg instanceof Integer) ? new BigDecimal((Integer) avg) : (BigDecimal) avg;
    return avgValue.doubleValue();
  }

  private void computeBaseMetrics(SensorContext context, List<InputFile> inputFiles) {
    for (InputFile groovyFile : inputFiles) {
      computeBaseMetrics(context, groovyFile);
    }
  }

  private void computeBaseMetrics(SensorContext context, InputFile groovyFile) {
    File file = groovyFile.file();
    if (file.exists()) {
      loc = 0;
      comments = 0;
      currentLine = 0;
      fileLinesContext = fileLinesContextFactory.createFor(groovyFile);
      Charset encoding = context.fileSystem().encoding();
      try (InputStreamReader streamReader = new InputStreamReader(new FileInputStream(file), encoding)) {
        List<String> lines = FileUtils.readLines(file, encoding);
        GroovyLexer groovyLexer = new GroovyLexer(streamReader);
        groovyLexer.setWhitespaceIncluded(true);
        TokenStream tokenStream = groovyLexer.plumb();
        Token token = tokenStream.nextToken();
        Token nextToken = tokenStream.nextToken();
        while (nextToken.getType() != Token.EOF_TYPE) {
          handleToken(token, nextToken.getLine(), lines);
          token = nextToken;
          nextToken = tokenStream.nextToken();
        }
        handleToken(token, nextToken.getLine(), lines);
        saveMetric(context, groovyFile, CoreMetrics.LINES, nextToken.getLine());
        saveMetric(context, groovyFile, CoreMetrics.NCLOC, loc);
        saveMetric(context, groovyFile, CoreMetrics.COMMENT_LINES, comments);
      } catch (TokenStreamException e) {
        LOG.error("Unexpected token when lexing file : " + file.getName(), e);
      } catch (IOException e) {
        LOG.error("Unable to read file: " + file.getName(), e);
      }
      fileLinesContext.save();
    }
  }

  private static void highlightFiles(SensorContext context, List<InputFile> inputFiles) {
    for (InputFile inputFile : inputFiles) {
      new GroovyHighlighterAndTokenizer(inputFile).processFile(context);
    }
  }

  private static <T extends Serializable> void saveMetric(SensorContext context, InputComponent inputComponent, Metric<T> metric, T value) {
    context.<T>newMeasure()
      .withValue(value)
      .forMetric(metric)
      .on(inputComponent)
      .save();
  }

  private void handleToken(Token token, int nextTokenLine, List<String> lines) {
    int tokenType = token.getType();
    int tokenLine = token.getLine();
    if (isComment(tokenType)) {
      if (isNotHeaderComment(tokenLine)) {
        comments += nextTokenLine - tokenLine + 1 - numberEmptyLines(token, lines);
      }
      for (int commentLineNb = tokenLine; commentLineNb <= nextTokenLine; commentLineNb++) {
        fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, commentLineNb, 1);
      }
    } else if (isNotWhitespace(tokenType) && tokenLine != currentLine) {
      loc++;
      fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, tokenLine, 1);
      currentLine = tokenLine;
    }
  }

  private int numberEmptyLines(Token token, List<String> lines) {
    List<String> relatedLines = getLinesFromToken(lines, (GroovySourceToken) token);
    long emptyLines = relatedLines.stream().map(String::trim).filter(EMPTY_COMMENT_LINES::contains).count();
    return (int) emptyLines;
  }

  private static List<String> getLinesFromToken(List<String> lines, GroovySourceToken gst) {
    List<String> newLines = new ArrayList<>(lines.subList(gst.getLine() - 1, gst.getLineLast()));

    int lastLineIndex = newLines.size() - 1;
    String lastLine = newLines.get(lastLineIndex).substring(0, gst.getColumnLast() - 1);
    newLines.set(lastLineIndex, lastLine);

    String firstLine = newLines.get(0).substring(gst.getColumn() - 1);
    newLines.set(0, firstLine);

    return newLines;
  }

  private boolean isNotHeaderComment(int tokenLine) {
    return !(tokenLine == 1 && settings.getBoolean(GroovyPlugin.IGNORE_HEADER_COMMENTS));
  }

  private static boolean isNotWhitespace(int tokenType) {
    return !(tokenType == GroovyTokenTypes.WS ||
      tokenType == GroovyTokenTypes.STRING_NL ||
      tokenType == GroovyTokenTypes.ONE_NL || tokenType == GroovyTokenTypes.NLS);
  }

  private static boolean isComment(int tokenType) {
    return tokenType == GroovyTokenTypes.SL_COMMENT || tokenType == GroovyTokenTypes.SH_COMMENT || tokenType == GroovyTokenTypes.ML_COMMENT;
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }

}
